src/modules/emote_menu/components/VirtualizedList.jsx

```diff
@@ -0,0 +1,76 @@
+import React, {useCallback, useEffect, useMemo, useRef, useState} from 'react';
+
+const style = (index, rowHeight) => ({
+  position: 'absolute',
+  top: `${index * rowHeight}px`,
+  height: `${rowHeight}px`,
+  width: '100%',
+});
+
+const headerStyle = (rowHeight) => ({
+  position: 'sticky',
+  top: '0px',
+  height: `${rowHeight}px`,
+  width: '100%',
+});
+
+function VirtualizedList(
+  {totalRows, rowHeight, renderRow, windowHeight, stickyRows = [], onHeaderChange = () => {}, ...restProps},
+  ref = null
+) {
+  const listHeight = useMemo(() => rowHeight * totalRows, [totalRows, rowHeight]);
+  const headers = useMemo(() => stickyRows.sort((a, b) => b - a), [stickyRows]);
+
+  const [data, setData] = useState({
+    header: null,
+    rows: [],
+  });
+
+  useEffect(() => {
+    onHeaderChange(data.header);
+  }, [data.header]);
+
+  const wrapperRef = ref || useRef(null);
+
+  const isInViewport = useCallback(() => {
+    const {scrollTop} = wrapperRef.current;
+    const scrollBottom = scrollTop + windowHeight;
+
+    const startIndex = Math.floor(scrollTop / rowHeight);
+    const endIndex = Math.min(totalRows - 1, Math.floor(scrollBottom / rowHeight));
+
+    const rowsVisible = [];
+
+    for (let i = startIndex; i <= endIndex; i++) {
+      rowsVisible.push(i);
```

Comment:
curious, why not use `slice` for this?

Reply:
What would we be slicing? we're pushing the index